src/storage/engine/mod.rs

```diff
@@ -48,27 +47,21 @@ pub trait Engine: Send + Sync + Debug {
     fn async_snapshot(&self, ctx: &Context, callback: Callback<Box<Snapshot>>) -> Result<()>;
 
     fn write(&self, ctx: &Context, batch: Vec<Modify>) -> Result<()> {
-        let finished = Event::new();
-        let finished2 = finished.clone();
         let timeout = Duration::from_secs(DEFAULT_TIMEOUT_SECS);
-
-        try!(self.async_write(ctx, batch, box move |res| finished2.set(res)));
-        if finished.wait_timeout(Some(timeout)) {
-            return finished.take().unwrap();
+        let event = async!(_cb, try!(self.async_write(ctx, batch, _cb)));
```

Comment:
@BusyJay Good idea! I also felt format is a little ugly, but was not able to come up with the better one.